src/main/java/org/ethereum/beacon/discovery/pipeline/info/FindNodeResponseHandler.java

```diff
@@ -3,30 +3,112 @@
+  private final Collection<Integer> distances;
   private int totalPackets = -1;
   private int receivedPackets = 0;
 
+  public FindNodeResponseHandler(final Collection<Integer> distances) {
+    this.distances = distances;
+  }
+
   @Override
-  public synchronized boolean handleResponseMessage(NodesMessage msg) {
+  public synchronized boolean handleResponseMessage(NodesMessage message, NodeSession session) {
     if (totalPackets == -1) {
-      totalPackets = msg.getTotal();
+      totalPackets = message.getTotal();
       if (totalPackets < 1 || totalPackets > MAX_TOTAL_PACKETS) {
         throw new RuntimeException("Invalid number of total packets: " + totalPackets);
       }
     } else {
-      if (totalPackets != msg.getTotal()) {
+      if (totalPackets != message.getTotal()) {
         throw new RuntimeException(
             "Total number differ in different packets for a single response: "
                 + totalPackets
                 + " != "
-                + msg.getTotal());
+                + message.getTotal());
       }
     }
     receivedPackets++;
+
+    // Parse node records
+    logger.trace(
+        () ->
+            String.format(
+                "Received %s node records in session %s. Total buckets expected: %s",
+                message.getNodeRecords().size(), session, message.getTotal()));
+    message.getNodeRecords().stream()
+        .filter(this::isValid)
+        .filter(record -> hasCorrectDistance(session, record))
+        .forEach(
+            nodeRecordV5 -> {
+              foundNodes.add(nodeRecordV5);
+              updateNodeRecord(session, nodeRecordV5);
+            });
+
     return receivedPackets >= totalPackets;
   }
+
+  public synchronized List<NodeRecord> getFoundNodes() {
+    return foundNodes;
+  }
+
+  private void updateNodeRecord(final NodeSession session, final NodeRecord nodeRecordV5) {
+    NodeRecordInfo nodeRecordInfo = NodeRecordInfo.createDefault(nodeRecordV5);
+    final NodeTable nodeTable = session.getNodeTable();
+    final Bytes nodeId = nodeRecordV5.getNodeId();
+    final Optional<NodeRecordInfo> existingRecord = nodeTable.getNode(nodeId);
+    if (isUpdateRequired(nodeRecordV5, existingRecord)) {
+      // Update node table with new node record
+      nodeTable.save(nodeRecordInfo);
+
+      if (session.getNodeId().equals(nodeId)) {
+        // Node sent us a new version of their own ENR, update the session.
+        session.updateNodeRecord(nodeRecordV5);
+      }
+    }
+  }
+
+  private boolean isUpdateRequired(
+      final NodeRecord newRecord, final Optional<NodeRecordInfo> existingRecord) {
+    return existingRecord.isEmpty()
+        || existingRecord.get().getNode().getSeq().compareTo(newRecord.getSeq()) < 0;
+  }
+
+  private boolean isValid(final NodeRecord record) {
+    if (!record.isValid()) {
+      logger.debug("Rejecting invalid node record {}", record);
+      return false;
+    }
+    return true;
```

Comment:
nit: I'd return `record.isValid()`, and use the condition just as a decision to debug...

Reply:
although that admittedly does introduce a chance to have a change in value, maybe that's why this was done this way..

Reply:
Calling `isValid` requires performing potentially expensive cryptographic operations, hence the single call.  Potentially validity could be cached but it currently isn't.

Reply:
true, but we can still call once by 
```
boolean result = record.isValid();
if (! result) ...

return isValid;
```